### DISCUSSION TITLE
Integration with solid-community-server

### DISCUSSION THREAD

**Discussion by franzzua** [m-ld/m-ld-spec#71] [Discussion]
[solid-commmunity-server](https://github.com/solid/community-server) is a modular implementation of solid server and i'm going to integrate it with m-ld-js for collaborative editing of rdf documents.
WebSocket protocol is not suitable for this, because it doesn't support headers, and webId-OIDC auth will not work properly. And other header-based solid specifications too.
I suppose that one can partly implement PubsubRemote through HTTP requests and default solid notification mechanism (which just send notification about changed resources without diffs or something else).
So architecture should be like options 2 and 3 in [document](https://github.com/m-ld/m-ld-spec/files/6579923/NGI.Pointer.-.attachment.pdf) 
* client must proceed authorization on server and get signed access token for a specific resource
* with this token client have access to a m-ld domain via WebRTC 
* server may implement STUN/TURN or signaling server for WebRTC
* client can send operations to server via PATCH 
* client can get changes from server since some clock time
* client can get current state from server via GET
* client can listen on notifications for specific resources

**Comment by gsvarovsky** [comment]
OK great, thanks for looking at this, @franzzua!

m-ld clones are fully peer-to-peer, and they have no understanding of any 'server', or even any clone that is more important than the others. Of course, it is no problem to use a server in the application architecture. You can also use one inside a `PubsubRemotes` subclass, to route messages. All the current subclasses do this, one way or another.

However, the PubsubRemotes base class is in charge of what messages are sent, and provides the subclass with payloads as Buffers. So the PubsubRemotes protocol can't be changed, only the transport. For example, it is PubsubRemotes that decides when to ask peers for current state (a 'snapshot') or a set of operations (a 'rev-up'), or to publish a change (an 'operation'). (You can see the main abstract protocol components in the [engine interfaces](https://github.com/m-ld/m-ld-js/blob/master/src/engine/index.ts).)

So:
> * client must proceed authorization on server and get signed access token for a specific resource

OK. This happens using pure Solid before you get access to the domain.

> * with this token client have access to a m-ld domain via WebRTC
> * server may implement STUN/TURN or signaling server for WebRTC

OK. You may have seen there is a [WebRTC helper class](https://github.com/m-ld/m-ld-js/blob/master/src/wrtc/WrtcPeering.ts), which is used by the [Ably remotes](https://github.com/m-ld/m-ld-js/blob/master/src/ably/AblyRemotes.ts). This only uses WebRTC for 'request/response' between two peers, not for publishing messages to all peers.

You could use WebRTC for publishing as well. You can decide whether to have a mesh (every clone has a connection to every other clone), or a hub-and-spoke model, in which the server forwards messages.

> * client can send operations to server via PATCH
> * client can get changes from server since some clock time
> * client can get current state from server via GET
> * client can listen on notifications for specific resources

Not necessarily OK. This is where the client seems to be trying to take control of the protocol. All of these functions are defined in the PubsubRemotes protocol, and the base class calls the subclass to implement the transport.

To do this more exactly, it would probably be better to replace PubsubRemotes completely with a new implementation of the abstract `MeldRemotes`.

However, one thing you cannot get away from, is that a Pod **must** have a **single** m-ld clone providing it with updates. You can't just use a m-ld `OperationMessage` as a patch directly to a Pod, because the messages must be processed by the CRDT to provide a convergent state. Once it is processed, it is possible to relay that state to the Pod using the updates generated by the clone engine.

So perhaps we are talking about **two** components here:
1. The 'remotes' implementation that relays messages between m-ld clones, and happens to use some server resources to do it (e.g. STUN/TURN and signalling).
2. A microservice, deployed near to the Pod (even inside its container), which has a m-ld clone and relays updates to the Pod using normal Solid HTTP. This can also relay changes coming the other way.

**Reply by franzzua** [reply]
Thanks for clarifications!

> two components here

Yes, and it's almost clear with first component. Just one thing: 
Is this method in charge of network topology? `present(): Observable<string>` For example. it should return all clients in room for a mesh and just one client for a star, is not it?

It's more interesting with server component and its form.
> a Pod must have a single m-ld clone

Yes, sure. But if some clients don't know about this feature and perform simple SPARQL update on a resource we should convert this update into operations and and put them into m-ld. So we need a module inside ResourceStore, that handles every update.

Also we can send operations from a client's clone to a server's clone in same manner via HTTP patch with `content-type: m-ld/update` for example. 

So, I imagine it like this:
![M-ld_solid](https://user-images.githubusercontent.com/3264482/129213611-33f0e48f-7433-421d-9132-f8bdf0b51e40.png)
At this scheme one client sends operations to a Solid POD, where specific handler wraps it to a patch object. M-ld module in a Resource Store handles every updates of m-ld resources through a local m-ld clone. 
Things to be implemented:
* `m-ld handler` inside solid and client code to talk with it.
* `m-ld store module`

Problems:
* There is two source of truth in Solid server: a resource and a m-ld clone associated with it. So every file change should be accepted by m-ld. For example, somebody replaces a resource with totally different content. It should not be an exceptional case for m-ld store module.

**Reply by gsvarovsky** [reply]
Wonderful!

> Is this method in charge of network topology? `present(): Observable<string>` For example. it should return all clients in room for a mesh and just one client for a star, is not it?

The `present` set is only used for clone "recovery", when a clone is brand new, or returns online, or notices that it has missed a message. It is not important for publishing of operations while online. So, it is not in charge of overall topology; only which peers are visible to help a clone recover.

So you are right, it could return all clones for a mesh and just one clone for a star. It could also be used to make different clones available at different places – for example, only clones that are geographically nearby.

---

Let's be clear that it is _not_ good to allow simple HTTP updates to a resource directly to the file, while that resource is shared with m-ld, because of the risk it will not converge with m-ld's state. Instead, you can proxy all the HTTP operations through the embedded m-ld clone. I'm pretty sure this is what you are suggesting anyway, I'm just checking.

So in your picture, all the arrows that go into the `Resource Store` are routed into the `m-ld module` (only for the resources where m-ld is active, of course), and the arrows going down to a `Resource` are coming from the `m-ld module`.

At the moment this means you have to translate SPARQL queries to m-ld's API. This should be okay for simple stuff, although not too pretty. However, we can make m-ld support SPARQL natively.

---

One problem is that changes to the resource that happen via simple HTTP must be broadcast out of the Pod to the rest of the domain.

I think that since we already have the `m-ld handler` inside the Pod, it would be good to make this use a websocket connection. Then there is no problem with traffic going in both directions.

---

Also, I don't think that we need to have a special `main client`, because every client can publish its updates to other clones **and** to the Pod `m-ld/update` end-point. (Note that the clone that is _inside_ the Pod counts as the **single** clone keeping the Pod in sync.) This has advantages:
- you don't need to maintain any state to know who is the 'main'. Even if the originator goes offline, everyone can keep working.
- every clone has its permissions checked, instead of just the main clone. This means it's not possible to accidentally have a clone for someone who should not have access.
- now, every clone is making a connection to the Pod. If this is a websocket as I suggest above, then maybe we don't even need WebRTC, because we can route all updates through the Pod.

**Reply by franzzua** [reply]
> proxy all the HTTP operations through the embedded m-ld clone. 

Yes, you are right.

> translate SPARQL queries to m-ld's API

Solid community server currently uses [SPARQLAlgebra](https://github.com/joachimvh/SPARQLAlgebra.js#readme) representation. It maybe difficult for translation or native support. I suppose that it may be implemented afterwards. 

> WebSocket

I think that it should not be main transport. Solid server already has WebSocket server, without authentication. So anybody can get unauthorized access to it. I can serve another WebSocket server on other port, ok. Also, WebSocket does not support headers for authentication. So I should use another authentication methods with one-time temporary tokens stored at server. I think it's enough rude changes to a server specification.

Instead of it, clients can get changes from a server via GET requests sometimes. They can send their CRDT clock to a server and get only unhandled changes. Also server can send his clock on every request in specific header, so client can check if there are any changes.

>  every client can publish its updates to the POD

Great. It really simplifies schema. Am I right, that operations should be published at least once? So could duplicates be a problem?

It seems that WebRTC will become best transport for communication between clients, it's enough fast and reliable, so I don`t see any reason for replacing it with something else.

**Reply by gsvarovsky** [reply]
> Solid community server currently uses [SPARQLAlgebra](https://github.com/joachimvh/SPARQLAlgebra.js#readme) representation

We're in luck! That's already used inside m-ld. So, all we need to do is provide an API method and wire it through. I did this recently for the RDF/JS `match` method.

> I think that [WebSocket] should not be main transport... Instead of it, clients can get changes from a server via GET requests sometimes.

OK, this should work!

> They can send their CRDT clock to a server and get only unhandled changes. Also server can send his clock on every request in specific header, so client can check if there are any changes.

This requires one extra method on the [`MeldLocal` interface](https://github.com/m-ld/m-ld-js/blob/814bfd833b38de19fdd7dc411bbaf7b5c2bad732/src/engine/index.ts#L188) to get the current clock time.

The server response can be generated using the `revupFrom` method. So the strategy looks good, and simple.

> Am I right, that operations should be published at least once? So could duplicates be a problem?

No – a clone uses the message clock time to check whether it has already received it. The only problem could be with network utilisation, but some accidental duplicate messages are not an issue.

I'm excited, this should work!

---

New question: if all HTTP methods are being proxied through m-ld anyway, then what is the file for? Why not just _only_ use m-ld for a "shared" resource?

**Reply by franzzua** [reply]
> Why not just only use m-ld for a "shared" resource?
Yes, we can store state in leveldb on the server side.

**Comment by gsvarovsky** [comment]
I will write some more documentation for `MeldRemotes` and `PubsubRemotes`, which should help this discussion 🙂

**Reply by gsvarovsky** [reply]
Done, on the [edge branch](https://github.com/m-ld/m-ld-js/tree/edge), e.g. [PubsubRemotes.ts](https://github.com/m-ld/m-ld-js/blob/edge/src/engine/remotes/PubsubRemotes.ts)

**Comment by gsvarovsky** [comment]
I have added SPARQL support to https://github.com/m-ld/m-ld-js/issues/74 (RDF/JS interface support in the JS engine), as discussed above https://github.com/m-ld/m-ld-spec/discussions/71#discussioncomment-1185224 (note this reply is hidden under Github's "View more" button).

**Reply by franzzua** [reply]
Wow, it's great!
I will try to use it slightly later

**Reply by gsvarovsky** [reply]
Haha sorry I haven't actually implemented it yet, just added the requirement to the ticket!

**Reply by gsvarovsky** [reply]
SPARQL query methods are implemented in the pre-release https://www.npmjs.com/package/@m-ld/m-ld/v/0.7.1-2

Documentation is on the edge sub-domain, e.g. https://edge.js.m-ld.org/interfaces/meldreadstate.html

The method signatures may change following conversation with the RDF/JS folks, see https://gitter.im/rdfjs/public?at=613f74e0f428f97a9f83ab55 (thread)

**Comment by renyuneyun** [comment]
Hi. I'm interested in using m-ld for a JS project, while hopefully uses Solid Pod as one of the storages (for, e.g., synchronization across devices while supporting off-line editing). I know RDF pretty well and have been involved with Solid for quite some time. However, for m-ld, after reading documents, I actually have several questions, including:
1. How to start developing using m-ld to save and/or manipulate my data (the [getting-started doc](https://edge.js.m-ld.org/#getting-started) doesn't look complete to me, at least by reading that page alone); and
2. How to use Solid as one node?

From the discussion above, does it mean a way is to use "SPARQL query methods for m-ld" when communicating with my Pod? Does it require any changes to the Solid service (e.g. do I need to use a custom Solid server implementation with added functionality)? Is this approach effectively equivalent to storing all states / transaction history into the Pod?

It would be really appreciated if any hints are suggested.

**Reply by gsvarovsky** [reply]
Hi Rui, welcome.

Apologies, the documentation for using RDF directly (via the RDF/JS API) is [incomplete](https://github.com/m-ld/m-ld-js/issues/179).  This is because the main API is JSON-LD, and direct RDF support is a later addition. There is a live code example (https://js.m-ld.org/#using-the-rdfjs-api) which should help.

Using Solid as one node requires that there is precisely one 'master' m-ld clone that is synchronised with Solid. There are options for how to do this:
1. you can use a custom Solid server with a m-ld clone embedded in it;
2. you can have a separate local process as a sidecar; or
3. you can elect one of the client clones to be the master.

These options are discussed in the [document](https://github.com/m-ld/m-ld-spec/files/6579923/NGI.Pointer.-.attachment.pdf) referenced above.

In all cases the master clone must serialise its state and history somewhere persistent. It would be theoretically possible to store all of this as RDF in the Solid Pod itself; however this option would require a new API layer in the m-ld JavaScript engine, which would be a lot of work. At the moment the engine insists that the persistence has an [abstract-level interface](https://github.com/Level/abstract-level) – in other words, the storage API is at the layer underneath RDF. Also, storing all the raw data as RDF would be very inefficient.
